File: app/src/main/java/com/teamunemployment/lolanalytics/front_page/BaseActivityPersistenceInteractor.java
package com.teamunemployment.lolanalytics.front_page;

import com.teamunemployment.lolanalytics.io.RESTApiExecutor;
import com.teamunemployment.lolanalytics.io.RealmExecutor;
import com.teamunemployment.lolanalytics.data.model.DoubleWrapper;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;


/**
 * @author Josiah Kendall
 * Interactor for fetching
 */

public class BaseActivityPersistenceInteractor {

    // Our repositories.
    private RealmExecutor realmExecutor;
    private RESTApiExecutor restApiExecutor;

    @Inject
    public BaseActivityPersistenceInteractor(RESTApiExecutor restApiExecutor, RealmExecutor realmExecutor) {
        this.restApiExecutor = restApiExecutor;
        this.realmExecutor = realmExecutor;
    }

    public void fetchWinRateForRole(long summonerId, String role, final BaseActivityContract.Presenter presenter) {

        Observable<DoubleWrapper> winRateObsservable = restApiExecutor.GetWinRateForRole(summonerId, role);
        winRateObsservable
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .map(doubleWrapper -> {
                    // realmExecutor. TODO - cache
                    return doubleWrapper;
                }).observeOn(AndroidSchedulers.mainThread()).subscribe(new Observer<DoubleWrapper>() {

            @Override
            public void onError(Throwable e) {
                presenter.handleError(e);
            }

            @Override
            public void onComplete() {

            }

            @Override
            public void onSubscribe(Disposable d) {

            }

            @Override
            public void onNext(DoubleWrapper doubleWrapper) {
                presenter.onWinRateLoaded(doubleWrapper.data);
            }

        });

    }
}

File: app/src/main/java/com/teamunemployment/lolanalytics/front_page/BaseActivityPersistenceInteractor.kt
package com.teamunemployment.lolanalytics.front_page

import co.metalab.asyncawait.async
import com.teamunemployment.lolanalytics.Utils.Network
import com.teamunemployment.lolanalytics.data.model.DoubleWrapper
import com.teamunemployment.lolanalytics.data.model.Result
import com.teamunemployment.lolanalytics.data.room.AppDatabase
import com.teamunemployment.lolanalytics.io.networking.RetrofitFactory


import retrofit2.Call
import retrofit2.Response


/**
 * @author Josiah Kendall
 * Interactor for fetching
 */

class BaseActivityPersistenceInteractor
constructor(private val retrofitFactory : RetrofitFactory, // Our repositories.
            private val database : AppDatabase,
            private val network : Network) {

    private lateinit var baseActivityService: BaseActivityService

    fun fetchWinRateForRole(summonerId : String, role : String, region : String, presenter: BaseActivityContract.Presenter) {
        async {
            val url = network.getUrlForRegion(region)
            baseActivityService = retrofitFactory.produceRetrofitInterface(BaseActivityService::class.java, url)
            val call : Call<Result<DoubleWrapper>> = baseActivityService.getWinRateForRole(summonerId, role)
            val response : Response<Result<DoubleWrapper>> = await { call.execute() }

            val result : Result<DoubleWrapper> = response.get()
            val code = result.resultCode // do something here?
            presenter.onWinRateLoaded(result.data.data)
        }.onError {
            presenter.showMessage("Offline - check you internet connection")

        }
    }

    private fun Response<Result<DoubleWrapper>>.get() : Result<DoubleWrapper> {
        if (body() != null) {
            return body()!!
        }

        return Result(404, DoubleWrapper())
    }
}
